pep-0692.rst

```diff
@@ -0,0 +1,344 @@
+
+Rationale
+=========
+
+PEP 589 introduced the ``TypedDict`` type constructor that supports dictionary types consisting of string keys and values of potentially different types. This makes ``TypedDict`` a perfect candidate to be used for more precise ``**kwargs`` typing. However, specifing ``**kwargs`` type using a ``TypedDict`` means, as mentioned earlier, that each keyword argument specified by ``**kwargs`` is a ``TypedDict`` itself. For instance:
+
+::
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+    
+    def foo(**kwargs: Movie) -> None: ...
+
+means that each keyword argument in ``foo`` is itself a ``Movie`` dictionary that has a ``name`` key with a string type value and a ``year`` key with an integer type value. Therefore, in order to support specifing ``kwargs`` type as a ``TypedDict`` without breaking current behaviour, a new syntax has to be introduced.
+
+Specification
+=============
+
+To support the aforementioned use case we propose to use the double asterisk syntax inside of the type annotation. The required grammar change is discussed in more detail in section `Grammar Changes`_. Continuing the previous example:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+would mean that the ``**kwargs`` comprise two keyword arguments specified by ``Movie`` (i.e. a ``name`` keyword of type ``str`` and a ``year`` keyword of type ``int``). Then, inside the function itself, appropriate type checking would take place:
+
+::
+
+    def foo(**kwargs: **Movie) -> None:
+        kwargs["name"].capitalize()      # OK!
+        kwargs["year"] + 42              # OK!
+        kwargs["name"] + kwargs["year"]  # WRONG! Unsupported operand types for + ("str" and "int")
+
+The ``kwargs`` parameter inside the function should be treated exactly the same way as any other ``TypedDict``.
+
+Function calls with standard dictionaries
+-----------------------------------------
+
+Calling a function that has ``**kwargs`` typed using the ``**kwargs: **Movie`` sytnax with a standard dictionary must raise an error. For example:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+    movie: dict[str, str] = {"name": "Life of Brian", "year": 1979}
```

Comment:
This line should also raise an error, because 1979 is not a str.
```suggestion
    movie: dict[str, object] = {"name": "Life of Brian", "year": 1979}
```